go/teams/box_audit.go

```diff
@@ -0,0 +1,1099 @@
+	}
+	if team == nil {
+		return nil, fmt.Errorf("got nil team from loader")
+	}
+
+	// If the team sigchain state was constructed with support for the
+	// merkleRoots map, the map will be non-nil but empty. It will only be nil
+	// if the state is cached from a team load before box summary hash support.
+	if team.chain().GetMerkleRoots() == nil {
+		if force {
+			return nil, fmt.Errorf("failed to get a non-nil merkleRoots map after full reload")
+		}
+		return loadTeamForBoxAuditInner(mctx, teamID, true)
+	}
+	return team, nil
+}
+
+type merkleCheckpoints map[keybase1.UserVersion]keybase1.Seqno
+
+func getAllPUKCheckpointsSinceMerkleSeqno(teamchain *TeamSigChainState, merkleSeqnoCheckpoint keybase1.Seqno) (merkleCheckpoints, error) {
+	checkpoints := make(merkleCheckpoints)
+	// We only check users currently in the team, which means we skip over any
+	// users, who for example, have reset (and possibly have added a new PUK),
+	// but have not been let back into the team by an admin.
+	for uv, logPoints := range teamchain.inner.UserLog {
+		logPoint := logPoints[len(logPoints)-1]
+		if logPoint.Role == keybase1.TeamRole_NONE {
+			continue
+		}
+		merkleSeqno := logPoint.SigMeta.PrevMerkleRootSigned.Seqno
+		latest := merkleSeqno
+		if latest < merkleSeqnoCheckpoint {
+			latest = merkleSeqnoCheckpoint
+		}
+		checkpoints[uv] = latest
+	}
+	return checkpoints, nil
+}
+
+// calculateCurrentSummary calculates the box summary as it is currently for
+// all users in the team (i.e., if the team were rotated right now, what the summary
+// should be afterwards).
+func calculateCurrentSummary(mctx libkb.MetaContext, team *Team) (summary *boxPublicSummary, err error) {
+	currentRoot, err := mctx.G().GetMerkleClient().FetchRootFromServer(mctx, 5*time.Minute)
+	if err != nil {
+		return nil, err
+	}
+	if currentRoot.Seqno() == nil {
+		return nil, fmt.Errorf("got nil current merkle root")
+	}
+	return calculateSummaryAtMerkleSeqno(mctx, team, currentRoot.Seqno())
+}
+
+// calculateChainSummary calculates the box summary as implied by the team sigchain and previous links,
+// using the last known rotation and subsequent additions as markers for PUK freshness.
+func calculateChainSummary(mctx libkb.MetaContext, team *Team) (summary *boxPublicSummary, err error) {
+	return calculateSummaryAtMerkleSeqno(mctx, team, nil)
+}
+
+// calculateSummaryAtMerkleSeqno calculates the summary at the merkleSeqno, and
+// if merkleSeqno is nil, it uses the merkle root at the beginning of the
+// team's (and its ancestors') generation.
+func calculateSummaryAtMerkleSeqno(mctx libkb.MetaContext, team *Team, merkleSeqno *keybase1.Seqno) (summary *boxPublicSummary, err error) {
+	defer mctx.TraceTimed(fmt.Sprintf("calculateSummaryAtMerkleSeqno(%s, %v)", team.ID, merkleSeqno), func() error { return err })()
+
+	if merkleSeqno == nil {
+		tmp, err := merkleSeqnoAtGenerationInception(mctx, team.chain())
+		if err != nil {
+			return nil, err
+		}
+		merkleSeqno = &tmp
+	}
```

Comment:
so instead of doing this here, change `merkleSeqno` to a value type, and do this calculation in `calculateChainSummary`

if this is the only call site, of course, maybe i missed something